### PR TITLE
feat: Add authentication logic

### PR DESCRIPTION
**Pull request by AlekseiLitkovetc**
The following things were done:

- Adding auth data to the local storage on SignIn or SingUp actions
- Removing auth data from the local storage on LogOut or session expiration
- Updating auth state in case of page refreshing
- Refreshing auth data in case of auth token was expired

### CODE REVIEW

File: src/axiosCromwell.ts [hunk 1]
```diff
@@ -1,10 +1,51 @@
...
+axiosCromwell.interceptors.request.use(
+  config => {
+    const accessToken = localStorage.getItem('accessToken');
+    if (accessToken) {
+      config.headers["Authorization"] = accessToken;
+    }
+    return config;
+  },
+  error => {
+    Promise.reject(error);
```
Comment (YBogomolov) [inline src/axiosCromwell.ts]: Probably, you wanted to return this? You can write this in a short form: `error => Promise.reject(error)`, or with eta-elimination — just `Promise.reject`.

File: src/axiosCromwell.ts [hunk 2]
```diff
@@ -1,10 +1,51 @@
...
+  }
+);
+
+axiosCromwell.interceptors.response.use(
+  response => {
+    return response;
+  },
+  error => {
+    const originalRequest = error.config;
+    let refreshToken = localStorage.getItem('refreshToken');
```
Comment (YBogomolov) [inline src/axiosCromwell.ts]: As you're not changing `refreshToken`, you can declare it as `const` as well

File: src/axiosCromwell.ts [hunk 3]
```diff
@@ -1,10 +1,51 @@
...
+    return config;
+  },
+  error => {
+    Promise.reject(error);
+  }
+);
+
+axiosCromwell.interceptors.response.use(
+  response => {
+    return response;
```
Comment (YBogomolov) [inline src/axiosCromwell.ts]: I'd added an identity function somewhere in the utils module ;)

File: src/axiosCromwell.ts [hunk 4]
```diff
@@ -1,10 +1,51 @@
...
+  },
+  error => {
+    const originalRequest = error.config;
+    let refreshToken = localStorage.getItem('refreshToken');
+    if (refreshToken && error.response.status === 401 && !originalRequest._retry) {
+      originalRequest._retry = true;
+      return axiosCromwell
+        .get('/auth/refresh', {params: {refreshToken: refreshToken}})
```
Comment (YBogomolov) [inline src/axiosCromwell.ts]:
```suggestion
        .get('/auth/refresh', {params: { refreshToken }})
```

File: src/axiosCromwell.ts [hunk 5]
```diff
@@ -1,10 +1,51 @@
...
+        .then(response => {
```
Comment (YBogomolov) [inline src/axiosCromwell.ts]: If you declare error handler as `async error =>  ...`, then you can `await` this axios call, and process in in a more linear fashion.

Comment (YBogomolov) [reply src/axiosCromwell.ts]: And if you do this, you can rewrite `return Promise.reject(error)` as just `throw error` instead.

File: src/store/auth/actions.ts [hunk 6]
```diff
@@ -1,31 +1,55 @@
...
+    const authHeaders: AuthHeaders = getAuthHeaders(response);
+    dispatch(signInSuccess(authHeaders));
+  }).catch(error => console.log(error))
 });
 
 export const signUp = (signUpData: SignUpData): CmwlThunkAction => (dispatch => {
-  dispatch(signUpSuccess());
-  // ToDo: Add axios async call
+  axiosCromwell.post('/auth/signUp', signUpData).then(response => {
+    const authHeaders: AuthHeaders = getAuthHeaders(response);
```
Comment (YBogomolov) [inline src/store/auth/actions.ts]: I'd suggest moving `: AuthHeaders` to the definition of `getAuthHeaders` function, and omit it here. It'd be more clean.

File: src/store/auth/actions.ts [hunk 7]
```diff
@@ -1,31 +1,55 @@
...
+
+export const authenticateUser = () => ({
+  type: AUTH_AUTHENTICATE_USER
 });
 
 export const signOut = (): AuthActionTypes => ({
   type: AUTH_SIGN_OUT
 });
+
+export const sessionExpired = (): AuthActionTypes => ({
```
Comment (YBogomolov) [inline src/store/auth/actions.ts]: What's the reasoning in making `sessionExpired`, `authenticateUser`, `signOut` lazy? Not as a nitpick, just out of curiosity ;)

**Review by YBogomolov — CHANGES_REQUESTED** [review]
Good job! I'd suggest making some minor changes to clean up the code.

File: src/axiosCromwell.ts (hunk 1 above)
Comment (AlekseiLitkovetc) [reply src/axiosCromwell.ts]: Thank you!

I was interested in finding more information about `eta-elimination` (just to be sure that I understood it correctly) but I haven't found any description. Could you please help me with it?

File: src/axiosCromwell.ts (hunk 2 above)
Comment (AlekseiLitkovetc) [reply src/axiosCromwell.ts]: You are totally correct :+1:

File: src/axiosCromwell.ts (hunk 3 above)
Comment (AlekseiLitkovetc) [reply src/axiosCromwell.ts]: Ahah, I also thought about it and about adding fp-ts for this purpose :smile: 
But I guess this case is not a reason for adding fp-ts (I'm going to make a separate request in the future where we will be using more functional constructs).

So yes, let's add our own identity function.

File: src/axiosCromwell.ts (hunk 4 above)
Comment (AlekseiLitkovetc) [reply src/axiosCromwell.ts]: :pray:
Done

File: src/axiosCromwell.ts (hunk 5 above)
Comment (AlekseiLitkovetc) [reply src/axiosCromwell.ts]: Brilliant :gem:

File: src/store/auth/actions.ts (hunk 6 above)
Comment (AlekseiLitkovetc) [reply src/store/auth/actions.ts]: Exactly! :+1: 
Done

File: src/store/auth/actions.ts (hunk 7 above)
Comment (AlekseiLitkovetc) [reply src/store/auth/actions.ts]: Shame on me :worried: 
Fixed

**Comment by AlekseiLitkovetc** [comment]
@YBogomolov thank you for your great review!

And I would be only glad if you nitpick to my requests :smile: because it would be additional motivation to take a look at some things deeper and growth faster :relieved:

Thank you :pray:

**Review by YBogomolov — APPROVED** [review]

**Pull request merged**